### ISSUE TITLE
Remove as much internal state from components as possible

### ISSUE THREAD

**Issue by taoeffect** [okTurtles/group-income#297]
### Problem

Having components keep track of their own state is considered poor practice for various reasons. The main reason though is that it makes the application difficult to reason about and maintain.

Components should be "purely functional" in the sense that all of their state is derived from the vuex store, computed properties based on that store, or `props` that are passed in to the component from a parent component.

There's only a handful of exceptions for stuff in the `data ()` section of a component:

- Bindings to form elements (e.g. for `vee-validate` to validate)
    + Place this type of data under the `form` key in `data ()`
- "Irrelevant" state for debug purposes, for example the `history` variable in `TimeTravel.vue`.
    + Place this type of data under the `ephemeral` key in `data ()` (see example below)
- Configuration for child components, for example the `sliderConfig` variable in `TimeTravel.vue` which provides the configuration for `<vue-slider>`
    + Place this type of data under the `config` key in `data ()`
- The state represents *ephemeral state we might not care to save beyond the life of the view unless some other action is performed*. A good example of this is the usage of `latestContractState` in `Join.vue`, which displays the group's profile details to the user, that aren't relevant if they choose not to join.
    + Edit: In this case the data should be **explicitly** marked as ephemeral by placing it under an `ephemeral` key in `data ()`:

    ```js
    // FILE: components/user-image.vue
    export default {
      name: 'UserImage',
      props: ['username'],
      async mounted () {
        if (!this.profile) {
          let userContractId = await namespace.lookup(this.username)
          let state = await latestContractState(userContractId)
          this.ephemeral.url = state.attributes.picture
        }
      },
      computed: {
        profile () {
          return this.$store.getters.memberProfile(this.username)
        },
        pictureURL () {
          return this.profile ? this.profile.globalProfile.picture : this.ephemeral.url
        }
      },
      data () {
        return {
          ephemeral: {
            url: null
          }
        }
      }
    }
    ```

### Solution

- Go through all of the code and rewrite as needed to conform to the above rules
- Remove as many variables from `data ()` as possible, for example almost all `error` variables should be removed and instead accessed through the variables that vee-validate [automatically injects](https://github.com/baianat/vee-validate#usage) on the component via `this.$validator.errorBag` and `this.errors`
- Convert anything that can be converted into a computed property or a passed-in `props` into one of those
- Codify the above rules into the Contributors Guidelines and/or wiki
- **There are a lot of .vues that need to have their `data ()` section put into a `form` property as per above rules**

/cc @ktamas77

**Comment by taoeffect** [comment]
Added `Documentation` tag and edited "Solution" to add need to put these rules into Contributor's guide and/or Wiki.

**Comment by taoeffect** [comment]
Updated issue with an example.

Also want to note that instead of using `this.errorMsg` type stuff, we should move everything over to the global error notification system when it's implemented (#175).

**Comment by taoeffect** [comment]
There is now a data model rules section in our documentation:

- https://github.com/okTurtles/group-income-simple/blob/master/docs/Style-Guide.md#group-income-data-model-rules

Not sure if we've cleaned up everything in the codebase but closing because stale (#456).
